### PR TITLE
Modules: Load the import map polyfill when needed

### PR DESCRIPTION
**Pull request by gziolo**
## What?

Explores whether there is a simple way to address the issue raised by @felixarntz in https://github.com/WordPress/gutenberg/issues/55942#issuecomment-1863602182:

> For the import maps polyfill, it looks like that JS file is always loaded? I was using latest Chrome and it loaded for me, which it probably shouldn't (since Chrome supports import maps). Is there a way to perform a quick support check with inline JS instead so that the polyfill can only be loaded as needed? It's over 12kB, so if we can avoid that for most sites, that would be a good win.

## Why?

It should be easier to handle in WordPress core, but as a quick experiment I replicated how we handle conditional polyfills in the past:

https://github.com/WordPress/wordpress-develop/blob/1bca32e30ede3f8731290b052b35c431d091e2f7/src/wp-includes/script-loader.php#L204-L214

## How?

I used the conditional check for feature detection included on MDN:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#feature_detection

## Screenshots or screencast 

It works correctly with browsers that support import maps. I tested with Safari and Chrome:

https://github.com/WordPress/gutenberg/assets/699132/3486fed3-4fc0-40f1-b98f-54d7d91f9b4a

### CODE REVIEW

**Comment by gziolo** [comment]
I only make it a normal pull request to be able to use Gutenberg PR Previewer powered by Playground. I'm not sure if that will help though.

File: lib/experimental/modules/class-gutenberg-modules.php [hunk 1]
```diff
@@ -123,15 +123,21 @@ public static function print_module_preloads() {
...
-		$import_map = self::get_import_map();
-		if ( ! empty( $import_map['imports'] ) ) {
-			wp_print_script_tag(
-				array(
-					'src'   => gutenberg_url( '/build/modules/importmap-polyfill.min.js' ),
-					'defer' => true,
-				)
-			);
-		}
+		$test = 'HTMLScriptElement.supports?.("importmap")';
```
Comment (felixarntz) [inline lib/experimental/modules/class-gutenberg-modules.php]: `?.` isn't well supported by older browsers as far as I know, so probably safer to avoid it here, given this is for a polyfill check. Maybe better like this:
```suggestion
		$test = 'HTMLScriptElement.supports && HTMLScriptElement.supports("importmap")';
```

File: lib/experimental/modules/class-gutenberg-modules.php [hunk 2]
```diff
@@ -215,4 +221,4 @@ function gutenberg_enqueue_module( $module_identifier ) {
 add_action( 'wp_head', array( 'Gutenberg_Modules', 'print_module_preloads' ) );
 
 // Prints the script that loads the import map polyfill in the footer.
-add_action( 'wp_footer', array( 'Gutenberg_Modules', 'print_import_map_polyfill' ), 11 );
+add_action( 'wp_head', array( 'Gutenberg_Modules', 'print_import_map_polyfill' ), 11 );
```
Comment (felixarntz) [inline lib/experimental/modules/class-gutenberg-modules.php:282]: Just curious what's the rationale from changing from `wp_footer` to `wp_head`? Is it because the other modules are printed in the head?

If so, wouldn't we want to also print it _before_ the module scripts (i.e. lower priority than `10`)?

**Review by felixarntz — COMMENTED** [review]
Thanks @gziolo, this looks great! Two small concerns below.

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 2 above)
Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php:282]: I mirrored WP core as there might be something specific in how `document.write` works. Once I confirm it works as expected in not supported browsers, I will play with where it should be located.

Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php:282]: @aduth knows best as he discovered this technique for polyfills.

Comment (felixarntz) [reply lib/experimental/modules/class-gutenberg-modules.php:282]: I'm merely wondering about the order of scripts. Isn't the polyfill needed _before_ the relevant `script[type="importmap"]` tags so that it works?

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 1 above)
Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php]: It seems to work as expected even with `?.` in Firefox 107 that doesn't support import maps:

https://caniuse.com/import-maps

The polyfill loads as expected. I see some entries on the console printer, so I need to confirm whether it actually works. Well, I see the interactivity script loaded so it's probably all good.

**Comment by gziolo** [comment]
It works as expected with browsers that don't support import maps, like Firefox 107:


https://github.com/WordPress/gutenberg/assets/699132/80a2fd73-6596-4760-9c9d-ddd7240134d8

@felixarntz, I'll look tomorrow into how to optimize it based on your feedback.

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 2 above)
Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php:282]: The order doesn't seem to matter. I don't know how much we can move it later in the HTML document:

Screenshot 2023-12-20 at 23 00 26

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 1 above)
Comment (felixarntz) [reply lib/experimental/modules/class-gutenberg-modules.php]: Firefox 107 is still not that old, relatively speaking. My point is `?.` isn't always supported, so since this is about a polyfill, my take is that we should make sure the code to check is _very_ widely compatible. It's a simple change that enhances compatibility.

Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php]: I planned to refactor it as you suggested. I simply didn't explain myself clearly enough. I was focused on testing last evening 😄

Comment (felixarntz) [reply lib/experimental/modules/class-gutenberg-modules.php]: Oh ok, thanks for clarifying!

**Review by luisherranz — APPROVED** [review]
Tested it in iOS, iPadOS and macOS versions with Safari 13, 14 and 15, and it works as expected.

I tested again the `document.head.appendChild(importMapPolyfill)` solution and that didn't work, so it seems like `document.write` is the key to making this work.

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 1 above)
Comment (felixarntz) [reply lib/experimental/modules/class-gutenberg-modules.php]: @gziolo What about this change? Looking at https://caniuse.com/mdn-javascript_operators_optional_chaining, the operator is certainly well supported, but still not everywhere, including Internet Explorer. I don't see a huge value in using it here, primarily because it is used for a polyfill check which has the sole purpose of cross-browser support, and it only reduces the code by a few characters anyway.

In other words: Strictly speaking, we don't need to support those browsers, but to me there is no benefit to _not_ doing so in this particular case.

Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php]: On my todo list. I didn’t find time to revisit the implementation but I fully agree with your rationale 👍👍👍

Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php]: I committed your comment. I hope it’s enough and I should really do it earlier 🙈

Comment (felixarntz) [reply lib/experimental/modules/class-gutenberg-modules.php]: @gziolo Sorry I didn't realize you simply hadn't come back to this yet - didn't mean to rush 🙃 

Thank you, that looks good to me!

**Review by felixarntz — APPROVED** [review]
Thanks @gziolo, LGTM!

File: lib/experimental/modules/class-gutenberg-modules.php (hunk 2 above)
Comment (gziolo) [reply lib/experimental/modules/class-gutenberg-modules.php:282]: I tested it again today, and it looks like it works also when included in the footer. However, I don't know how much it's related to how Interactivity API works.

**Comment by Mamaduka** [comment]
I think this PR affected `Largest Contentful Paint (classic)` and `LCP - TTFB (classic)` stats - https://www.codevitals.run/project/gutenberg.

**Comment by luisherranz** [comment]
That's weird, because even though adding scripts with `document.write` is blocking and synchronous, it should not affect the latest versions of the browsers, which I guess is what the CWV is using to test.

We could try adding `async`, but if LCP is affected, it probably means the polyfill is being loaded when it shouldn't.

Previous related conversations:

- https://github.com/WordPress/gutenberg/pull/56143#issuecomment-1834150263
- https://github.com/WordPress/gutenberg/pull/56699#issuecomment-1834451359

**Comment by Mamaduka** [comment]
@luisherranz, this could be due to the WP hook changing from `wp_footer` to `wp_head`, but I've not tested this theory.

**Comment by luisherranz** [comment]
Yeah, but theoretically, in modern browsers, all it adds is a `script` tag with an `if` that resolves to `false`:

```html
<script>
  if ( ! ( HTMLScriptElement.supports && HTMLScriptElement.supports( "importmap" ) ) ) {
    // Nothing else to do, because the if is always false in all the modern browsers.
  }
</script>
```

So that shouldn't affect the LCP even if it's in the `head`.

**Comment by luisherranz** [comment]
I'm not sure I'm looking in the correct place, but if I am, it seems like the performance tests are using the latest chromium version:

https://github.com/WordPress/gutenberg/blob/52ff927c9d495f212f2687234a352ae6646c87c6/bin/plugin/commands/performance.js#L213

Theoretically, that version should not load the polyfill.

**Comment by Mamaduka** [comment]
That's correct; perf tests use the latest browser versions provided for Playwright - https://github.com/microsoft/playwright/releases/tag/v1.39.0.

**Comment by luisherranz** [comment]
Tested Chromium 119 installed and run via `playwright` 1.3.9 locally in a arm Mac and confirmed that it does not load the polyfill.

Screenshot 2024-01-09 at 14 05 40

**Comment by luisherranz** [comment]
Just to be sure, we are looking at these numbers, right?

https://github.com/WordPress/gutenberg/actions/runs/7409967406/job/20161251805?pr=57256


1. `front-end-block-theme`

| (index)               | f7afa64d69f90ac4494c3a0b49b73d7fe9f673ee | trunk      |
|-----------------------|------------------------------------------|------------|
| timeToFirstByte       | '51.8 ms'                               | '51.3 ms'  |
| largestContentfulPaint| '106.2 ms' | '104.4 ms' |
| lcpMinusTtfb          | '52.8 ms'                               | '52.5 ms'  |

2. `front-end-classic-theme`

| (index)               | f7afa64d69f90ac4494c3a0b49b73d7fe9f673ee | trunk      |
|-----------------------|------------------------------------------|------------|
| timeToFirstByte       | '39.15 ms'                              | '38.35 ms' |
| largestContentfulPaint| '113.8 ms' | '103.85 ms' |
| lcpMinusTtfb          | '73.9 ms'                               | '64.3 ms'  |

---

Maybe we should run them a few more times to see if the numbers stay still?

Any other ideas about what might be going on here?

cc: @youknowriad, @gziolo

**Comment by gziolo** [comment]
That’s surprising results. It’s also unclear to me why it would impact negatively only the classic theme.

**Comment by gziolo** [comment]
We could try moving the polyfill to the footer as it seems to work in my basic testing, but as Luis pointed out it's a simple JS check for over 90% of users.

**Comment by luisherranz** [comment]
> Maybe we should run them a few more times to see if the numbers stay still?

Now that we've got a few more PRs after this one, it's clear that it affected the LCP (https://www.codevitals.run/project/gutenberg):

Screenshot 2024-01-13 at 11 57 47

---

As it still doesn't make sense that a single script with an `if` that evaluates to `false` is causing a performance issue, one test we could do to see if the browser that is used in the CI supports import maps or not is replacing the `document.write()` with a very expensive and blocking JS call. If that clearly increases the LCP, it means the `if` is not evaluating to `false`.
